File: seven-framework-server/internal/app/authorization/domain/access_explain.go
package domain

import "time"

type AccessUserRecord struct {
	UserID   int64  `db:"user_id"`
	Username string `db:"username"`
	Status   int    `db:"status"`
}

type AccessRoleSourceRecord struct {
	RoleID           int64  `db:"role_id"`
	RoleCode         string `db:"role_code"`
	RoleName         string `db:"role_name"`
	RoleStatus       int    `db:"role_status"`
	RoleDataScope    int    `db:"role_data_scope"`
	RoleSystemKey    string `db:"role_system_key"`
	AssignmentSource string `db:"assignment_source"`
	PostID           int64  `db:"post_id"`
	PostCode         string `db:"post_code"`
	PostName         string `db:"post_name"`
	PostDeptID       int64  `db:"post_dept_id"`
	PostOrgID        int64  `db:"post_org_id"`
}

type AccessGrantRecord struct {
	PermissionID     int64      `db:"permission_id"`
	PermissionCode   string     `db:"permission_code"`
	PermissionName   string     `db:"permission_name"`
	PermissionStatus int        `db:"permission_status"`
	FeatureCode      string     `db:"feature_code"`
	GrantSource      string     `db:"grant_source"`
	RoleID           int64      `db:"role_id"`
	RoleCode         string     `db:"role_code"`
	RoleName         string     `db:"role_name"`
	RoleStatus       int        `db:"role_status"`
	AssignmentSource string     `db:"assignment_source"`
	PostID           int64      `db:"post_id"`
	PostCode         string     `db:"post_code"`
	PostName         string     `db:"post_name"`
	PostDeptID       int64      `db:"post_dept_id"`
	PostOrgID        int64      `db:"post_org_id"`
	MenuID           int64      `db:"menu_id"`
	MenuName         string     `db:"menu_name"`
	MenuStatus       int        `db:"menu_status"`
	GrantedBy        int64      `db:"granted_by"`
	Source           string     `db:"source"`
	PermissionType   int        `db:"permission_type"`
	ExpireAt         *time.Time `db:"expire_at"`
}

type AccessRoleDeptRecord struct {
	RoleID int64 `db:"role_id"`
	DeptID int64 `db:"dept_id"`
}

type AccessMembershipRecord struct {
	Kind      string `db:"kind"`
	ID        int64  `db:"id"`
	OrgID     int64  `db:"org_id"`
	Hierarchy string `db:"hierarchy"`
}

File: seven-framework-web/src/services/admin/accessExplain.ts
import { request } from '@/api/request';

export type ApiIdentifier = number | string;
export type AccessDecision = 'ALLOW' | 'DENY';
export type PermissionGrantSource = 'ROLE_DIRECT' | 'MENU_DERIVED' | 'TEMPORARY';
export type RoleAssignmentSource = 'DIRECT_USER' | 'POST';

interface ApiPayload<T> {
  code: number;
  data: T;
  message?: string;
}

export interface AccessPostSource {
  postId: ApiIdentifier;
  postCode: string;
  postName: string;
  deptId: ApiIdentifier;
  orgId: ApiIdentifier;
}

export interface AccessRoleSource {
  roleId: ApiIdentifier;
  roleCode: string;
  roleName: string;
  roleStatus: number;
  declaredDataScope: string;
  roleAssignmentSource: RoleAssignmentSource;
  post?: AccessPostSource;
}

export interface DataScopeContributor {
  roleId: ApiIdentifier;
  roleCode: string;
  declaredScopeType: string;
  winning: boolean;
  deptIds: ApiIdentifier[];
}

export interface EffectiveDataScope {
  userId: ApiIdentifier;
  scopeType: string;
  deptIds: ApiIdentifier[];
  orgIds: ApiIdentifier[];
  contributors: DataScopeContributor[];
}

export interface PermissionGrantChain {
  permissionGrantSource: PermissionGrantSource;
  roleId?: ApiIdentifier;
  roleCode?: string;
  roleName?: string;
  roleAssignmentSource?: RoleAssignmentSource;
  post?: AccessPostSource;
  menuId?: ApiIdentifier;
  menuName?: string;
  menuPath?: string;
  grantedBy?: ApiIdentifier;
  source?: string;
  expireAt?: string;
  active: boolean;
  reasonCode: string;
}

export interface EffectivePermission {
  permissionId: ApiIdentifier;
  permissionCode: string;
  permissionName: string;
  effective: boolean;
  featureCode?: string;
  featureEnabled: boolean;
  grants: PermissionGrantChain[];
}

export interface EffectiveAccess {
  userId: ApiIdentifier;
  username: string;
  status: number;
  authorizationRoot: boolean;
  roleSources: AccessRoleSource[];
  dataScope: EffectiveDataScope;
  permissionSummary: {
    effectiveCount: number;
    filteredCount: number;
    temporaryCount: number;
  };
  permissions: {
    current: number;
    size: number;
    total: number;
    records: EffectivePermission[];
  };
}

export interface PermissionExplanation {
  userId: ApiIdentifier;
  permissionCode: string;
  decision: AccessDecision;
  reasonCode: string;
  matchedPermissionCodes: string[];
  chains: PermissionGrantChain[];
  feature?: { code: string; enabled: boolean };
  evaluatedAt: string;
}

export interface EffectiveAccessParams {
  current?: number;
  size?: number;
  keyword?: string;
  sourceType?: string;
  effective?: boolean;
}

export async function fetchEffectiveAccess(userId: ApiIdentifier, params: EffectiveAccessParams) {
  const response = await request<ApiPayload<EffectiveAccess>>(
    `/api/system/user/${encodeURIComponent(String(userId))}/effective-access`,
    { method: 'GET', params },
  );
  const data = response.data;
  return {
    ...data,
    roleSources: data.roleSources ?? [],
    dataScope: {
      ...data.dataScope,
      deptIds: data.dataScope?.deptIds ?? [],
      orgIds: data.dataScope?.orgIds ?? [],
      contributors: data.dataScope?.contributors ?? [],
    },
    permissionSummary: {
      effectiveCount: Number(data.permissionSummary?.effectiveCount ?? 0),
      filteredCount: Number(data.permissionSummary?.filteredCount ?? 0),
      temporaryCount: Number(data.permissionSummary?.temporaryCount ?? 0),
    },
    permissions: {
      current: Number(data.permissions?.current ?? 1),
      size: Number(data.permissions?.size ?? params.size ?? 20),
      total: Number(data.permissions?.total ?? 0),
      records: data.permissions?.records ?? [],
    },
  } satisfies EffectiveAccess;
}

export async function fetchPermissionExplanation(userId: ApiIdentifier, permissionCode: string) {
  const response = await request<ApiPayload<PermissionExplanation>>(
    `/api/system/user/${encodeURIComponent(String(userId))}/access-explain`,
    { method: 'GET', params: { permissionCode } },
  );
  return {
    ...response.data,
    matchedPermissionCodes: response.data.matchedPermissionCodes ?? [],
    chains: response.data.chains ?? [],
  } satisfies PermissionExplanation;
}
